### DISCUSSION TITLE
3.1.0に関する不具合や質問はこちら

### DISCUSSION THREAD

**Discussion by dongaba** [dongaba/TVerRec#93] [Issues]
### 3.1.0 Added

- マイページからの番組取得に対応
  - 設定ファイルに`$script:myPlatformUID`を追加
  - 設定ファイルに`$script:myPlatformToken`を追加
- 新規キーワードの追加
  - `mypage/fav` = マイページ > お気に入り (TVerが管理する情報を取得)
  - `mypage/favorite` = マイページ > お気に入り (TVerRecがお気に入り番組、出演者、特集から各エピソードに変換)
  - `mypage/later` = マイページ > あとでみる
  - `mypage/resume` = マイページ > 続きから再生
  - `tag/independence` = 独立局
  - `tag/texttrackon` = 字幕あり
  - `tag/short` = 短尺(10分以内)
  - `tag/trailer` = 予告
- Chrome拡張機能のTVerRec Assistantを同梱
  - [使い方はWikiのこちらのページ参照](https://github.com/dongaba/TVerRec/wiki/platform_uid%E3%81%A8platform_token%E3%81%AE%E5%8F%96%E5%BE%97#chrome%E6%8B%A1%E5%BC%B5%E6%A9%9F%E8%83%BD%E3%81%AEtverrec-assistant%E3%82%92%E4%BD%BF%E3%81%86)
- トップページで急上昇バラエティを処理可能に
- ダウンロードリスト作成処理でリストファイルにあるリンクを除外するオプションを追加
  - 設定ファイルに`$script:listGenHistoryCheck`を追加
- Windows 環境で GUI 版の実行ログ表示について最大行数を設定可能に
  - 設定ファイルに`$script:guiMaxExecLogLines`を追加
- ダウンロードリスト作成処理でマルチスレッド処理を可能に(数倍の高速化)
- WSBの日本語化用スクリプトの追加
  - [使い方はWikiのこちらのページ参照](https://github.com/dongaba/TVerRec/wiki/Windows%E3%82%B5%E3%83%B3%E3%83%89%E3%83%9C%E3%83%83%E3%82%AF%E3%82%B9) 

### 3.1.0 Changed

- メッセージ出力フォーマットの変更
- ファイルロックの精度を改善(ファイルロックのメカニズムを変更)
- エラーハンドリングを改善
- TVerからの情報取得処理を正規化・高速化
- トップページのパース精度向上
- マルチスレッド処理のデフォルトスレッド数を10に変更(ダウンロードリスト作成処理のマルチスレッド対応に伴うAPIアクセス集中の抑制)
- 連続稼働時のメモリ使用量の削減
- リストファイル出力時にキーワードファイルのコメントを削除
- DockerコンテナのBaseImageをUbuntu:24.04に変更
- キーワードサンプルファイルで`tag/short`のみを有効化(キーワードファイル未設定でも動作イメージがわかりやすいように)
- Dockerfileを最新のコードに合わせて修正
- GUIから設定をする際の高度な設定を複数タブに分割
- Mac用のffmpegをVersion7に対応
- Mac用のffmpegでApple Silicon対応

### 3.1.0 Fixed

- `&amp;`が正しく置換されない不具合を修正
- Unix 時間への変換にタイムゾーンが考慮されていなかった不具合を修正
- トップページの処理でシリーズやシーズンを取得できていなかった不具合を修正
- リストファイルのサンプルファイルのレイアウトが間違っていた問題を修正
- WSB用のスクリプトが壊れていた問題を修正
- GUIからの設定の際に`$script:ytdlNonTVerFileName`の保存が正しくできなかった問題を修正
- GUIから設定の際に設定値に`=`を含む場合に正しく読み込まれたなかった問題を修正

## What's Changed
* update workflows by @dongaba in https://github.com/dongaba/TVerRec/pull/70
* update by @dongaba in https://github.com/dongaba/TVerRec/pull/71
* merge master to dev by @dongaba in https://github.com/dongaba/TVerRec/pull/72
* Merge branch 'master' of github.com:dongaba/TVerRec by @dongaba in https://github.com/dongaba/TVerRec/pull/74
* Merge dev branch as of 2024/04/29 to beta by @dongaba in https://github.com/dongaba/TVerRec/pull/75
* 3.1.0 beta 1 by @dongaba in https://github.com/dongaba/TVerRec/pull/76
* 3.1.0 release ready from dev to beta by @dongaba in https://github.com/dongaba/TVerRec/pull/77
* 3.1.0 release ready by @dongaba in https://github.com/dongaba/TVerRec/pull/78
* remove unnecessary step by @dongaba in https://github.com/dongaba/TVerRec/pull/79
* Beta to Dev by @dongaba in https://github.com/dongaba/TVerRec/pull/80
* sync by @dongaba in https://github.com/dongaba/TVerRec/pull/82
* 「Tverサイト以外のベースファイル名」が設定保存時に全体がシングルクオートされない問題を修正 by @dongaba in https://github.com/dongaba/TVerRec/pull/85
* Dockerfile fix by @dongaba in https://github.com/dongaba/TVerRec/pull/87
* change tag to tag/short by @dongaba in https://github.com/dongaba/TVerRec/pull/89
* few more features by @dongaba in https://github.com/dongaba/TVerRec/pull/92

**Full Changelog**: https://github.com/dongaba/TVerRec/compare/v3.0.2...v3.1.0

<hr /><em>This discussion was created from the release <a href='https://github.com/dongaba/TVerRec/releases/tag/v3.1.0'>3.1.0</a>.</em>

**Comment by catherine-newman** [comment]
こちらのページのcomposeファイル例では/mnt/Videoと書いてありますが、現在/mnt/Saveになっているのでこの例で設定すると移動先ディレクトリを設定してもうまくいかないです。
https://github.com/dongaba/TVerRec/wiki/Docker%E3%82%B3%E3%83%B3%E3%83%86%E3%83%8A

**Reply by dongaba** [reply]
ご指摘ありがとうございます、更新が漏れていたようです。更新しました。

**Comment by icg51379** [comment]
Ver3.1.0 から使わせて頂いています。
GUI の リスト作成 と リストダウンロードを使用しています。

platform_uidとplatform_token を取得し、下記のように設定ファイルに記載しているのですが、
リスト作成をクリックしても、"GUI 実行ログ"のようにマイリストの情報を取得してきてくれません。
なにか、他に設定すべき点などありますでしょうか？
ブラウザの開発者ツール、Chrome拡張機能のTVerRec Assistantの両方で取得を試み同じ値が取得されることは確認済みです。
series の指定は期待通り動作します。

 -- user_setting.ps1 --
$script:myPlatformUID = '取得した文字列'
$script:myPlatformToken = '取得した文字列'

 -- GUI 実行ログ --
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
mypage/fav #マイページ > お気に入り (こちらの方が高速だが、どのように抽出されているか不明。多少多かったり少なかったりする)
　処理対象0本　処理済0本

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
mypage/favorite #マイページ > お気に入り
　処理対象0本　処理済0本

**Reply by dongaba** [reply]
こちらについてはちょっと原因がわかりません。
私はTVerIDは登録したことがないのですが、TVerIDを登録して使っていたりしますか?
もしTVerIDを登録しているのであれば、TVerIDを登録しない状態でplatform_uidとplatform_token を取得して試してみていただけないでしょうか? (TVerIDが原因かはわかりませんが。。)

**Reply by icg51379** [reply]
TVerIDを登録して使っています。TVerIDを登録しない状態でplatform_uidとplatform_token を取得して試してみます。結果は別途お知らせ致します。

**Reply by icg51379** [reply]
ChromeでTVerからログアウトした状態でplatform_uidとplatform_token を取得したら値が変わりました。user_setting.ps1を変更し、リスト作成 をクリックすると、マイページ登録した番組が期待通り list.csvに 追加されました。

**Reply by dongaba** [reply]
ご確認ありがとうございます。TVerIDにログインした状態だとうまく動かないようですね、今後の課題としておきます。
申し訳ありませんが、マイページ登録した番組をダウンロードする際はTVerIDにログインせずにご利用ください。

**Comment by icg51379** [comment]
Ver3.1.0 から使わせて頂いています。
動画容量を少なくするために、下記の設定をすると半数くらいの動画ダウンロードが失敗してしまいます。
$script:downloadWorkDir のフォルダに、番組名.jpg と 番組名.fhls-508-0.mp4 のような音声なしの動画ファイルが残されます。

-- user_setting.ps1 --
$script:ytdlOption = '-f bestvideo[height<=480]+bestaudio/best[height<=480]'

この事象は youtube-dl の動作に依存しており、dongaba 様では対処ができない事象になりますでしょうか？

**Reply by dongaba** [reply]
すみません、オプションの指定方法を少し直す必要があるようです。
`$script:ytdlOption = '-f "bestvideo[height<=480]+bestaudio/best[height<=480]"'`
の様に「"」を2箇所足してみていただけますか。

**Reply by icg51379** [reply]
「"」を2箇所足して試してみます。結果は別途お知らせ致します。

**Reply by icg51379** [reply]
$script:ytdlOption = '-f "bestvideo[height<=480]+bestaudio/best[height<=480]"'
としましたが、事象は解消しませんでした。

例えば私の環境では下記の動画ダウンロードが失敗してしまいます。
$script:downloadWorkDir のフォルダに、番組名.jpg と 番組名.fhls-508-0.mp4 のような音声なしの動画ファイルが残されます。
https://tver.jp/episodes/ep4o7ae0fl
https://tver.jp/episodes/ep5lsc8zjr
https://tver.jp/episodes/epyy7uir5n
https://tver.jp/episodes/epts2ol7xr
https://tver.jp/episodes/epy2phuef9
https://tver.jp/episodes/epsdinr93h
https://tver.jp/episodes/epy8uq3s2o

ただし、history.csv から該当ファイルの履歴を削除し再度ダウンロードを試みると、どれか一つくらいが成功したりします。

**Reply by dongaba** [reply]
当方の環境でも確認しましたが、同様の事象は発生するものの、そこまでの失敗率ではないようです。
何度か再ダウンロードするとダウンロードが成功することもある、ということはオプション指定が完全に間違っているわけではないということなので、ネットワークが切断されてしまったりしてダウンロードが中断されることがあるということだと思います。

抜本的な解決は難しいのですが、オプションを以下のように変えることで多少は問題が改善するかもしれません。
`$script:ytdlOption = '-f "best[height<=480] / bestvideo[height<=480]+bestaudio"'`
今までのオプションは映像のみのファイルと音声のみのファイルを別々にダウンロードして結合するような指定だったのですが、上記を使用することで映像と音声を含むファイルを優先するようになります。
結果としてダウンロード時間の短縮が期待できるためネットワーク断による影響が減るのではないかと思っています。

抜本的な解決にはなりませんが、`b.delete_trash.cmd`または`不要ファイル削除`を実行することでダウンロードが中断して`$script:downloadWorkDir`フォルダに残ってしまったファイルは削除できます。
また、`c.validate_video.cmd`または`番組整合性検証`を実行することでダウンロードが正常に終わっていない場合は履歴ファイルから該当の動画を削除しますので、再度ダウンロードを試すのが容易になります。
これらを全て自動でやるのが`start_tverrec.cmd`、または`ループ実行`です。

**Reply by icg51379** [reply]
再現性の確認等ありがとうございます。教えて頂いたOptionに変更して試してみます。それでも改善しない場合は $script:ytdlOption は設定しないで使わせて頂こうと思います。ご対応ありがとうございました。

**Reply by icg51379** [reply]
$script:ytdlOption = '-f "best[height<=480] / bestvideo[height<=480]+bestaudio"'
に変更してから、動画ダウンロードは失敗しなくなりました。
ご教示いただきありがとうございました。

**Comment by catherine-newman** [comment]
度々もうしわけございません。
一番最近のdockerイメージではこのエラーが出ています。
start_tverrec.sh: line 17: /usr/bin/pwsh: Permission denied
Finished ...

**Reply by dongaba** [reply]
最新のPowerShellで配布ファイルの実行権限がなくなったことが問題のようです。
強制的に実行権限を付与するように変更しました。
3.2.0とlatestでは修正済みです。

**Reply by catherine-newman** [reply]
ありがとうございます！
